### DISCUSSION TITLE
How to add to field if id already exists

### DISCUSSION THREAD

**Discussion by gaganbiswas** [pocketbase/pocketbase#4855] [Q&A] [answered]
I have a `stock` and a `virtual_stock` collection. I want to add data to the `virtual_stock` wherever my stock is added. I wrote this piece of code for the same which is working fine.

```js
routerAdd("POST", "/stock/add", (c) => {
  const info = $apis.requestInfo(c);
  const { data } = info.data;

  const stock_collection = $app.dao().findCollectionByNameOrId("stock");
  const vstock_collection = $app
    .dao()
    .findCollectionByNameOrId("virtual_stock");

  $app.dao().runInTransaction((txDao) => {
    for (let item of data) {
      const stock_record = new Record(stock_collection);
      const vstock_record = new Record(vstock_collection);

      const stock_form = new RecordUpsertForm($app, stock_record);

      stock_form.setDao(txDao); // use the transaction dao for the db operations
      stock_form.loadData(item);
      stock_form.submit();

      const vstock_form = new RecordUpsertForm($app, vstock_record);

      vstock_form.setDao(txDao); // use the transaction dao for the db operations
      vstock_form.loadData(item);
      vstock_form.submit();

      //   verify that the inserted record can be accessed by the current request client, otherwise rollback
      if (
        !txDao.canAccessRecord(
          stock_record,
          info,
          stock_collection.createRule
        ) &&
        !txDao.canAccessRecord(
          vstock_record,
          info,
          vstock_collection.createRule
        )
      ) {
        throw new ForbiddenError("The operation is not allowed.");
      }
    }
  });

  return c.json(200, { status: 1 });
});
```

But, I also wanna check if my stock & virtual stock already have the product and if it does then I just wanna update the quantity for the same.

My stock & virtual_stock has the same model:
id(string), product(relation), quantity(number) 

_Side Note: The data that I am sending is in the form of {product: string, quantity: number}[]_

**Comment by ganigeorgiev** [comment]
You can call `findRecordById` as part of your transaction and try catch the result, something like:

```js
let stock_record;
try {
    stock_record = txDao.findRecordById(stock_collection.id, item.id);
} catch {
    stock_record = new Record(stock_collection);
}

let vstock_record;
try {
    vstock_record = txDao.findRecordById(vstock_collection.id, item.id);
} catch {
    vstock_record = new Record(vstock_collection);
}
```

**Reply by ganigeorgiev** [reply, marked as answer]
On second read and based on your side-note, you want to check for `stock_record`/`vstock_record` based on the "product" field and not on the "id". In this case you can replace `findRecordById` in the above with `findFirstRecordByData(yourCollecitonNameOrId, "product", item.product)` (assuming that the `product` field is unique).

**Reply by gaganbiswas** [reply]
Thank you for the swift reply and solution.

**Reply by gaganbiswas** [reply]
So, I got the first part ok. But how will I update the quantity field if the product already exists. It says Value must be unique when I am trying to update using this:

```js
const stock_form = new RecordUpsertForm($app, stock_record);
stock_form.setDao(txDao); // use the transaction dao for the db operations
stock_form.loadData(item);
stock_form.submit();
```

**Reply by gaganbiswas** [reply]
Did something like this: 

```js
const stock_form = new RecordUpsertForm($app, stock_record);
stock_form.setDao(txDao);
stock_form.loadData({
product: stock_record.get("product") || item.product,
quantity: stock_record.getInt("quantity")
          ? stock_record.getInt("quantity") + item.quantity
          : item.quantity,
      });
stock_form.submit();
```

Getting the same result when doing this. And on the side note, yes my product field is unique.

**Reply by ganigeorgiev** [reply]
It is not clear what is the issue and what you are trying to do.

If you are receiving a "unique" validation error then either the `stock_record` is a new record with existing unique field values or you are not properly fetching the old record.

But in any case without details about your collections structure I don't think I'll be able to help you. If you are still not able to resolve your issue, please elaborate more clearly what you are trying to do and export your collections configuration as json (_Admin UI > Settings > Export collections_).

**Reply by ganigeorgiev** [reply]
Additionally make sure that the error is actually from the `submit` and not from somewhere else (_you can try to wrap it with `try/catch`_).

You may also want to change the `txDao.canAccessRecord` condition to `||` instead of `&&` if the requirement is both to be accessible (aka. throw an error if at least one of the `canAccessRecord` checks fail).

**Reply by gaganbiswas** [reply]
[pb_schema.json](https://github.com/pocketbase/pocketbase/files/15192239/pb_schema.json)

Let me explain properly what I am trying to achieve.

Say, I already have a record in the `stock` collection with the `product` field value 'abc' and quantity field value as 10.
Now, when adding data through the above route, one of the item in the data array has the the product field set to 'abc' and quantity as 30. In that case I want to update the 'abc' field value in my `stock` to 10 + 30 = 40.

I hope this helps in understanding. 

In more concise way, I want to update the `quantity` field of an existing product in the `stock` (if the product exists) else create a new record in the `stock` collection.

_P.S. The data array looks something like this:
[
{product: 'abc', quantity: 30},
{product: 'xyz', quantity: 10},
...
]_

**Reply by ganigeorgiev** [reply]
In that case I don't see why the above doesn't work.

You can try to add print statements to debug the failing part, for example:

```js
console.log("product:", item.product)
console.log("quantity:", item.quantity)

let stock_record;
try {
    stock_record = txDao.findFirstRecordByData(stock_collection.id, "product", item.product);
} catch {
    console.log("fallback to a new record - missing record with product", item.product)
    stock_record = new Record(stock_collection);
}

const stock_form = new RecordUpsertForm($app, stock_record);

stock_form.setDao(txDao)

stock_form.loadData({
    // or: "quantity+": item.quantity,
    quantity: stock_record.getInt("quantity") + item.quantity,
});

try {
    stock_form.submit();
} catch (err)  {
    console.log("failed to submit:", err)
}
```

**Reply by gaganbiswas** [reply]
Don't know what was wrong. Just copy pasted this and it worked. Made some tweaks and the final code works like charm. Thanks again!
This is the final code and it works as intended.

```js
routerAdd("POST", "/stock/add", (c) => {
  const info = $apis.requestInfo(c);
  const { data } = info.data;

  const stock_collection = $app.dao().findCollectionByNameOrId("stock");
  const vstock_collection = $app
    .dao()
    .findCollectionByNameOrId("virtual_stock");

  $app.dao().runInTransaction((txDao) => {
    for (let item of data) {
      let stock_record;
      try {
        stock_record = txDao.findFirstRecordByData(
          stock_collection.id,
          "product",
          item.product
        );
      } catch {
        stock_record = new Record(stock_collection);
      }

      const stock_form = new RecordUpsertForm($app, stock_record);
      stock_form.setDao(txDao);

      if (stock_record.get("id"))
        stock_form.loadData({
          "quantity+": item.quantity,
        });
      else stock_form.loadData(item);
      stock_form.submit();

      // Virtual Stock
      let vstock_record;
      try {
        vstock_record = txDao.findFirstRecordByData(
          vstock_collection.id,
          "product",
          item.product
        );
      } catch {
        vstock_record = new Record(vstock_collection);
      }

      const vstock_form = new RecordUpsertForm($app, vstock_record);
      vstock_form.setDao(txDao);

      if (vstock_record.get("id")) {
        vstock_form.loadData({
          "quantity+": item.quantity,
        });
      } else {
        vstock_form.loadData(item);
      }

      vstock_form.submit();
    }
  });

  return c.json(200, { status: 1 });
});
```
